extensions/security/deployment/src/main/java/io/quarkus/security/deployment/SecurityProcessor.java

```diff
@@ -85,6 +109,147 @@ void registerSecurityInterceptors(BuildProducer<InterceptorBindingRegistrarBuild
+            ctor.writeArrayValue(result, i, ctor.load(parameters.get(i).toString()));
+        }
+
+        return result;
+    }
+
+    private Map<MethodInfo, AnnotationInstance> gatherSecurityAnnotations(ApplicationIndexBuildItem indexBuildItem,
+            BeanRegistrar.RegistrationContext registrationContext) {
+        Set<DotName> securityAnnotations = SecurityAnnotationsRegistrar.SECURITY_BINDINGS.keySet();
+        AnnotationStore annotationStore = registrationContext.get(BuildExtension.Key.ANNOTATION_STORE);
+        Set<ClassInfo> classesWithSecurity = new HashSet<>();
+
+        Collection<ClassInfo> classes = indexBuildItem.getIndex().getKnownClasses();
+        for (ClassInfo classInfo : classes) {
+            boolean hasSecurityAnnotations = annotationStore.hasAnyAnnotation(classInfo, securityAnnotations);
+            if (!hasSecurityAnnotations) {
+                for (MethodInfo method : classInfo.methods()) {
+                    if (annotationStore.hasAnyAnnotation(method, securityAnnotations)) {
+                        hasSecurityAnnotations = true;
+                        break;
+                    }
+                }
+            }
+            if (hasSecurityAnnotations) {
+                classesWithSecurity.add(classInfo);
+            }
+        }
+
+        return gatherSecurityAnnotations(securityAnnotations,
+                classesWithSecurity, annotationStore);
+    }
+
+    private Map<MethodInfo, AnnotationInstance> gatherSecurityAnnotations(Set<DotName> securityAnnotations,
+            Set<ClassInfo> classesWithSecurity,
+            AnnotationStore annotationStore) {
+        Map<MethodInfo, AnnotationInstance> methodAnnotations = new HashMap<>();
+        for (ClassInfo classInfo : classesWithSecurity) {
+            Collection<AnnotationInstance> classAnnotations = annotationStore.getAnnotations(classInfo);
+            AnnotationInstance classLevelAnnotation = getSingle(classAnnotations, securityAnnotations);
+
+            for (MethodInfo method : classInfo.methods()) {
+                AnnotationInstance methodAnnotation = getSingle(annotationStore.getAnnotations(method), securityAnnotations);
+                methodAnnotation = methodAnnotation == null ? classLevelAnnotation : methodAnnotation;
+                if (methodAnnotation != null) {
+                    methodAnnotations.put(method, methodAnnotation);
+                }
+            }
+        }
+        return methodAnnotations;
+    }
+
+    private AnnotationInstance getSingle(Collection<AnnotationInstance> classAnnotations, Set<DotName> securityAnnotations) {
+        AnnotationInstance result = null;
+        for (AnnotationInstance annotation : classAnnotations) {
+            if (securityAnnotations.contains(annotation.name())) {
+                if (result != null) {
+                    throw new IllegalStateException("Duplicate security annotations on class " + annotation.target());
```

Comment:
I think the error message can be improved. `Duplicate` isn't the correct word here IMHO because a developer could do:

```
    @RolesAllowed("admin")
    @DenyAll
    public void doSomething() {
    }
```

and get this error

Reply:
yeah, and I thought duplicate covers that. Would be grateful for some tips on new name

Reply:
I would just say that multiple security annotations are present

Reply:
done